source/common/common/singleton.h

```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+/**
+ * Singleton pattern.
+ */
+template <class T> class Singleton {
+public:
+  /**
+   * Obtain an instance of the singleton for class T.
+   * @return T& a reference to the singleton for class T.
+   */
+  static T& get() {
```

Comment:
I wonder if we should return `const T&` here to avoid oopses if this template gets more use. I think in general that is what people will want. (I think all uses of Headers should compile with returning this).